File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0, head);
        ListNode segmentPrevious = dummy;

        while (true) {
            ListNode segmentTail = getKthNode(segmentPrevious, k);
            if (segmentTail == null) {
                break;
            }

            ListNode nextSegmentHead = segmentTail.next;

            ListNode previous = nextSegmentHead;
            ListNode current = segmentPrevious.next;
            while (current != nextSegmentHead) {
                ListNode next = current.next;
                current.next = previous;
                previous = current;
                current = next;
            }

            ListNode oldGroupStart = segmentPrevious.next;
            segmentPrevious.next = segmentTail;
            segmentPrevious = oldGroupStart;
        }
        
        return dummy.next;
    }

    private ListNode getKthNode(ListNode head, int k) {
        while (head != null && k > 0) {
            head = head.next;
            k--;
        }
        return head;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        val dummy = ListNode(0)
        dummy.next = head
        var fast: ListNode? = dummy
        var offset = n
        while (fast != null && offset >= 0) {
            fast = fast.next
            offset--
        }
        var slow: ListNode? = dummy
        while (fast != null) {
            fast = fast.next
            slow = slow?.next
        }

        slow?.next = slow?.next?.next
 
        return dummy?.next
    }
}
